sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs

```diff
@@ -220,6 +263,121 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+                )?;
+                ensure!(
+                    read_count > 0,
+                    TransportError::CommunicationError("Truncated GPIO response".to_string())
+                );
+                bytecount += read_count;
+            }
+            ensure!(
+                resp.encapsulation_header == Self::CMSIS_DAP_CUSTOM_COMMAND_GPIO,
+                TransportError::CommunicationError(
+                    "Unrecognized CMSIS-DAP response to GPIO request".to_string()
+                )
+            );
+            match resp.status {
+                0 => (),
+                n => bail!(TransportError::CommunicationError(format!(
+                    "HyperDebug error: {}",
+                    n
+                ))),
+            }
+            let skip_bytes = resp.struct_size as usize - (GPIO_MONITORIN_HEADER_SIZE - 1);
+
+            let mut databytes: Vec<u8> = Vec::new();
+            databytes.extend_from_slice(&resp.data[..bytecount - GPIO_MONITORIN_HEADER_SIZE]);
+
+            while databytes.len() < skip_bytes + resp.transcript_size as usize {
+                let original_length = databytes.len();
+                databytes.resize(original_length + 64, 0u8);
+                let c = self.inner.usb_device.borrow().read_bulk(
+                    cmsis_interface.in_endpoint,
+                    &mut databytes[original_length..],
+                )?;
+                databytes.resize(original_length + c, 0u8);
+            }
+
+            // Figure out how many of the low bits are used for storing the index of the signal
+            // hanving changed. (If only one signal, no bits are used, if two signals, then one
+            // bit is used, if three or four, then two bits are used, etc.)
+            let mut signal_bits = 0;
+            while (pin_names.len() - 1) >> signal_bits != 0 {
+                signal_bits += 1;
+            }
+            let signal_mask = (1u64 << signal_bits) - 1;
+
+            let mut cur_time: u64 = resp.start_timestamp;
+            let mut cur_level = Vec::new();
+            for i in 0..pin_names.len() {
+                cur_level.push(resp.start_levels & (1u16 << i) != 0);
+            }
+            let mut events = Vec::new();
+            let mut idx = skip_bytes;
+
+            // Now decode the list of events, each consisting of a variable legnth encoded 64-bit
+            // integer.
+            while idx < databytes.len() {
+                // Read 7 bits from each byte, least significant byte first.  High bit of one
+                // indicates more bytes belong to the same value.
+                let mut value = 0u64;
+                let mut shift = 0;
+                loop {
+                    value |= ((databytes[idx] & 0x7F) as u64) << shift;
+                    shift += 7;
+                    idx += 1;
+                    if (databytes[idx - 1] & 0x80) == 0 {
+                        break;
+                    }
+                }
```

Comment:
I was not aware that this was a standard format, with an existing crate for decoding.  Rather than add to dependencies, I have factored out the leb128 decoding, and used that name, so that readers can immediately recognize the name and format.